flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/sink/SocketClientSink.java

```diff
@@ -73,13 +84,44 @@ public void intializeConnection() {
 	 *			The incoming data
 	 */
 	@Override
-	public void invoke(IN value) {
+	public void invoke(IN value) throws Exception {
 		byte[] msg = schema.serialize(value);
 		try {
 			dataOutputStream.write(msg);
 		} catch (IOException e) {
-			throw new RuntimeException("Cannot send message " + value.toString() +
-					" to socket server at " + hostName + ":" + port, e);
+			retries = 0;
+			boolean success = false;
+			while ((retries < maxRetry || retryForever) && !success && isRunning){
+				try {
+
+					if (dataOutputStream != null) {
+						dataOutputStream.close();
+					}
+
+					if (client != null && !client.isClosed()) {
+						client.close();
+					}
+
+					if (!retryForever){
+						retries++;
+					}
+
+					client = new Socket(hostName, port);
+					dataOutputStream = new DataOutputStream(client.getOutputStream());
+					dataOutputStream.write(msg);
+					success = true;
+
+				}catch(Exception ee){
```

Comment:
Well, it actually depends on the exceptions for which we want to restart. If I'm not mistaken, then `new Socket()` and `dataOutputStream.write` only throw `IOExceptions`. Thus, we should change it.